File: poc/src/engine/deck.ts
import type { Card } from "./card";
import { standardDeck } from "./card";

export class Deck {
  cards: Card[];

  constructor(cards: Card[] = standardDeck()) {
    this.cards = [...cards];
  }

  shuffle(rng: () => number = Math.random): void {
    for (let i = this.cards.length - 1; i > 0; i--) {
      const j = Math.floor(rng() * (i + 1));
      [this.cards[i], this.cards[j]] = [this.cards[j], this.cards[i]];
    }
  }

  draw(n = 1): Card[] {
    if (n > this.cards.length) {
      throw new Error(`Cannot draw ${n} cards; only ${this.cards.length} remain`);
    }
    const drawn = this.cards.slice(0, n);
    this.cards = this.cards.slice(n);
    return drawn;
  }

  get remaining(): number {
    return this.cards.length;
  }
}

File: prototype/quintet/deck.py
"""Deck operations."""

from __future__ import annotations

import random
from dataclasses import dataclass, field

from quintet.card import Card, Rank, Suit


def standard_deck() -> list[Card]:
    return [Card(rank=r, suit=s) for s in Suit for r in Rank]


@dataclass
class Deck:
    cards: list[Card] = field(default_factory=standard_deck)

    def shuffle(self, rng: random.Random | None = None) -> None:
        rng = rng or random.Random()
        rng.shuffle(self.cards)

    def draw(self, n: int = 1) -> list[Card]:
        if n > len(self.cards):
            raise ValueError(f"Cannot draw {n} cards; only {len(self.cards)} remain")
        drawn = self.cards[:n]
        self.cards = self.cards[n:]
        return drawn

    def __len__(self) -> int:
        return len(self.cards)
